src/App.tsx

```diff
@@ -27,59 +28,53 @@ function App() {
   const [walletNotifyResults, setWalletNotifyResults] = useState<any>(null);
   const [walletSessionChangedResults, setWalletSessionChangedResults] =
     useState<any>(null);
-  const [currentSession, setCurrentSession] = useState<any>(null);
   const [extensionId, setExtensionId] = useState<string>('');
   const [invokeMethodRequests, setInvokeMethodRequests] = useState<
     Record<string, string>
   >({});
 
+  // Use useEffect to handle provider initialization and cleanup
   useEffect(() => {
-    console.log('extensionId', extensionId);
-    if (extensionId) {
-      provider.connect(extensionId);
+    let newProvider: Provider;
+    if (providerType === 'mock') {
+      newProvider = makeProvider(() => createSessionResult);
+    } else {
+      newProvider = new MetaMaskMultichainProvider();
+      if (extensionId) {
+        newProvider.connect(extensionId);
+      }
     }
+
+    setProvider(newProvider);
+
     return () => {
-      provider.disconnect();
+      newProvider?.disconnect();
     };
-  }, [extensionId]);
+  }, [providerType, createSessionResult, extensionId]);
 
   useEffect(() => {
-    const interval = setInterval(() => {
-      if (!currentSession) {
-        return;
+    if (extensionId && provider) {
+      provider.connect(extensionId);
+    }
+    return () => {
+      if (provider) {
+        provider.disconnect();
       }
-      setWalletSessionChangedResults({
-        jsonrpc: '2.0',
-        method: 'wallet_sessionChanged',
-        params: currentSession,
-      });
-    }, 1000);
-    return () => clearInterval(interval);
-  }, [currentSession]);
+    };
+  }, [extensionId, provider]);
```

Comment:
this seems to overlap with the useEffect above which also triggers on extensionId change

Reply:
ahh i see. maybe this only needs to handle the disconnect if the provider changes and then the useEffect above this doesn't need to do its own cleanup